cpp/src/io/parquet/reader_impl_chunking.cu

```diff
@@ -19,45 +19,60 @@
 
 #include <cudf/detail/iterator.cuh>
 #include <cudf/detail/utilities/integer_utils.hpp>
+#include <cudf/detail/utilities/vector_factories.hpp>
 
+#include <io/comp/nvcomp_adapter.hpp>
+
+#include <io/utilities/config_utils.hpp>
 #include <io/utilities/time_utils.cuh>
 
 #include <rmm/exec_policy.hpp>
 
 #include <thrust/binary_search.h>
 #include <thrust/iterator/constant_iterator.h>
 #include <thrust/iterator/discard_iterator.h>
+#include <thrust/logical.h>
 #include <thrust/sort.h>
+#include <thrust/unique.h>
+
+#include <cuda/functional>
+
+#include <numeric>
 
 namespace cudf::io::parquet::detail {
 
 namespace {
 
-struct cumulative_row_info {
-  size_t row_count;   // cumulative row count
-  size_t size_bytes;  // cumulative size in bytes
-  int key;            // schema index
+struct split_info {
+  row_range rows;
+  int64_t split_pos;
 };
 
+// the minimum amount of memory we can safely expect to be enough to
+// do a subpass decode. if the difference between the user specified limit and
+// the actual memory used for compressed/temp data is > than this value, we will still use
+// at least this many additional bytes.
```

Comment:
I don't undestand this sentence. Can you please give an example?

Reply:
What it's saying is:  
- Let's say the user has passed in a limit of 1GB
- We have somehow consumed 900 MB of that before we even start decompressing
- In theory we should only use 100 MB more memory, based on the user's limit.
- But we want to at least work with 200 MB at a time (to keep the GPU busy, etc), so it's just going to use 200 instead of the 100.

The counter-example is: 

- Let's say the user has passed in a limit of 1GB
- We have consumed 100 MB prior to decompression.
- We've got 800 MB left, which is more than 200, so we will decompress in 800 MB batches.

Reply:
I can add the example:
```
// Example:
// - user has specified 1 GB limit
// - we have read in 900 MB of compressed data
// - that leaves us 100 MB of space for decompression batches
// - to keep the gpu busy, we really don't want to do less than 200 MB at a time so we're just going to use 200 MB of space  even if that goes past the user-specified limit.
```

Reply:
the example is very clear, would not mind it